### PR TITLE
Warmboot Vlan neigh restore fix

### PR DESCRIPTION
**Pull request by prsunny**
**What I did**
1. During Warmboot, the restore_neighbor script sends out ARP/NS for Vlan interfaces based on **oper** status. Since Vlan interface is bound to bridge, it is up by default. Modified to wait for Vlan members to be added.

2. Logger is changed to syslog for getting correct timestamps for events

3. Nbrmgrd  push to kernel must happen only after warmboot neighbor restoration

**Why I did it**
To fix neigh restore issue for Vlans

**How I verified it**

**Details if related**

```
Aug 28 02:29:09.296187 str-msn2700-04 INFO swss#supervisord: start.sh restore_neighbors: started
Aug 28 02:29:22.277583 str-msn2700-04 INFO swss#restore_neighbor: Error: [Errno 2] No such file or directory: '/sys/class/net/Vlan1000/carrier'
Aug 28 02:29:33.573648 str-msn2700-04 NOTICE swss#vlanmgrd: :- doVlanTask: Add Vlan  1000
Aug 28 02:29:51.640572 str-msn2700-04 NOTICE swss#vlanmgrd: :- doVlanMemberTask: Add Vlan member:Ethernet4 to Vlan 1000
Aug 28 02:29:53.343235 str-msn2700-04 INFO swss#restore_neighbor: intf Vlan1000 is up
Aug 28 02:29:55.155619 str-msn2700-04 INFO swss#restore_neighbor: Add neighbor entries: family: IPv4, intf_idx: 45, ip: 192.168.0.2, mac: 00:00:00:11:22:33
Aug 28 02:29:55.155863 str-msn2700-04 INFO swss#restore_neighbor: Sending Neigh with family: IPv4, intf_idx: 45, ip: 192.168.0.2, mac: 00:00:00:11:22:33
Aug 28 02:30:01.209086 str-msn2700-04 NOTICE swss#orchagent: :- setHostIntfsOperStatus: Set operation status UP to host interface Ethernet96
Aug 28 02:30:02.681734 str-msn2700-04 NOTICE swss#vlanmgrd: :- doVlanMemberTask: Add Vlan member:Ethernet96 to Vlan 1000
Aug 28 02:30:03.952226 str-msn2700-04 NOTICE swss#portsyncd: :- main: PortInitDone
Aug 28 02:30:03.961584 str-msn2700-04 NOTICE swss#orchagent: :- addVlan: Create an empty VLAN Vlan1000 vid:1000
Aug 28 02:30:03.963923 str-msn2700-04 NOTICE swss#orchagent: :- addNeighbor: Created neighbor 00:00:00:11:22:33 on Vlan1000
Aug 28 02:30:03.963923 str-msn2700-04 NOTICE swss#orchagent: :- addNextHop: Created next hop 192.168.0.2 on Vlan1000
Aug 28 02:30:08.902206 str-msn2700-04 NOTICE swss#orchagent: :- addVlanMember: Add member Ethernet4 to VLAN Vlan1000 vid:1000 pid1000000000549
Aug 28 02:30:08.916187 str-msn2700-04 NOTICE swss#orchagent: :- addVlanMember: Add member Ethernet96 to VLAN Vlan1000 vid:1000 pid1000000000252

```

### CODE REVIEW

File: neighsyncd/restore_neighbors.py [hunk 1]
```diff
@@ -58,12 +76,29 @@ def is_intf_oper_state_up(intf):
...
     if state == '1':
         return True
     return False
 
+def is_intf_up(intf):
+    if not is_intf_oper_state_up(intf):
+         return False
+    if 'Vlan' in intf:
+        db = swsssdk.SonicV2Connector(host='127.0.0.1')
+        db.connect(db.STATE_DB, False)
```
[the remark this answers is not in the document]
Comment (zhenggen-xu) [reply neighsyncd/restore_neighbors.py]: yes, the program should need open the db connection once.

File: neighsyncd/restore_neighbors.py [hunk 2]
```diff
@@ -58,12 +76,29 @@ def is_intf_oper_state_up(intf):
...
+        table_name = 'VLAN_MEMBER_TABLE|{}|*'.format(intf)
+        key = db.keys(db.STATE_DB, table_name)
+        if key is None:
+            log_info ("Vlan member is not yet created")
+            return False
+        if is_intf_up.counter == 0:
```
[the remark this answers is not in the document]
Comment (zhenggen-xu) [reply neighsyncd/restore_neighbors.py]: See below comment. keep per interface counter will not block other interfaces, and each cycle (5 seconds for now), you have a chance to check interface status. so you won't wait in a sequence.

File: neighsyncd/restore_neighbors.py [hunk 3]
```diff
@@ -233,24 +271,22 @@ def restore_update_kernel_neighbors(intf_neigh_map, timeout=DEF_TIME_OUT):
 
 def main():
 
-    print "restore_neighbors service is started"
-
+    log_info ("restore_neighbors service is started")
```
[the remark this answers is not in the document]
Comment (zhenggen-xu) [reply neighsyncd/restore_neighbors.py]: I saw the code in the beginning:
"
logger = logging.getLogger(__name__)
logger.setLevel(logging.WARNING)
logger.addHandler(logging.NullHandler()
"
Was it to set the log level?

File: neighsyncd/restore_neighbors.py [hunk 4]
```diff
@@ -25,11 +25,29 @@
...
+import syslog
 
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.WARNING)
 logger.addHandler(logging.NullHandler())
 
+SYSLOG_IDENTIFIER = 'restore_neighbor'
+
+def log_info(msg):
+    syslog.openlog(SYSLOG_IDENTIFIER)
```
[the remark this answers is not in the document]
Comment (zhenggen-xu) [reply neighsyncd/restore_neighbors.py]: Let's change it so we won't give bad examples to future programmers. :)

File: cfgmgr/nbrmgrd.cpp [hunk 5]
```diff
@@ -50,6 +54,21 @@ int main(int argc, char **argv)
...
+        while (!nbrmgr.isNeighRestoreDone())
+        {
+            chrono::duration<double> time_span = chrono::duration_cast<chrono::duration<double>>
+                                                 (chrono::steady_clock::now() - starttime);
+            int pasttime = int(time_span.count());
+            SWSS_LOG_INFO("Kernel neighbor table restoration waited for %d seconds", pasttime);
+            if (pasttime > RESTORE_NEIGH_WAIT_TIME_OUT)
+            {
+                SWSS_LOG_WARN("Kernel neighbor table restore is not finished!");
+                break;
```
[the remark this answers is not in the document]
Comment (zhenggen-xu) [reply cfgmgr/nbrmgrd.cpp]: Hmm, Why do we need push neighbors to kernel after NeighRestoreDone?  If we push earlier, what will happen?

Also, should we only check the isNeighRestoreDone() when warmrestart is enabled?  If it is a cold boot/restart, we don't check this at all.

File: neighsyncd/restore_neighbors.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (prsunny) [reply neighsyncd/restore_neighbors.py]: Done

File: neighsyncd/restore_neighbors.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (prsunny) [reply neighsyncd/restore_neighbors.py]: Based on the group review, the original plan was to not have per interface counter. The intention was to wait till the interfaces are oper and prefer not to add further delays. Per interface counter seems to be slightly an overhead IMO.

File: neighsyncd/restore_neighbors.py [hunk 6]
```diff
@@ -58,12 +76,29 @@ def is_intf_oper_state_up(intf):
...
+    if 'Vlan' in intf:
+        db = swsssdk.SonicV2Connector(host='127.0.0.1')
+        db.connect(db.STATE_DB, False)
+        table_name = 'VLAN_MEMBER_TABLE|{}|*'.format(intf)
+        key = db.keys(db.STATE_DB, table_name)
+        if key is None:
+            log_info ("Vlan member is not yet created")
+            return False
+        if is_intf_up.counter == 0:
+            time.sleep(3*CHECK_INTERVAL)
```
[the remark this answers is not in the document]
Comment (prsunny) [reply neighsyncd/restore_neighbors.py]: Please check above

File: neighsyncd/restore_neighbors.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (prsunny) [reply neighsyncd/restore_neighbors.py]: Yes

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
retest this please

**Comment by tylerlinp** [comment]
nbrmgrd waits restore neighbors for 120s in normal startup?  isNeighRestoreDone true only if really do restore. I found VS tests about neighbor/nexthop (vrf new cases) failed because nbrmgrd cannot work.

**Comment by prsunny** [comment]
In normal startup, there is no wait as warmboot flag is disabled and the isNeighRestoreDone flag would be set without any wait. can you provide any logs that nbrmgrd is stuck in VS?

**Comment by tylerlinp** [comment]
nbrmgrd is stuck in VS because:
1. In VS startup.sh, now there is no start restore_neighbors.
2. In restore_neighbors.py, if not warmstart.isWarmStart() missing set_statedb_neigh_restore_done().

**Pull request merged**